src/twisted/internet/test/test_unix.py

```diff
@@ -429,18 +444,91 @@ def _dataReceived(self, data):
 
         dataToSend = b'some data needs to be sent'
         fdsToSend = [sendSocket.fileno(), recvSocket.fileno()]
-        ancillary = [(SOL_SOCKET, SCM_RIGHTS, pack('ii', *fdsToSend))]
+        ancillary, expectedCount = ancillaryPacker(fdsToSend)
         sendmsg(sendSocket, dataToSend, ancillary)
 
         receiver.doRead()
 
         # Verify that fileDescriptorReceived was called twice.
-        self.assertEqual(len(proto.fds), 2)
+        self.assertEqual(len(proto.fds), expectedCount)
 
         # Verify that received FDs are different from the sent ones.
         self.assertFalse(set(fdsToSend).intersection(set(proto.fds)))
```

Comment:
Just pushed a commit with the first approach I commented above:
* Bind sockets to known addresses.
* Have `FakeProtocol` track not only FDs but also `fstat`'s (`st_dev`, `st_ino`) tuples.
* Do sendmsg / doRead.
* Verify the sent (`st_dev`, `st_ino)` tuples against the ones tracked by `FakeProtocol`.

Feedback welcome.